### PR TITLE
Improve multidetection and add some calibration plots

### PR DESCRIPTION
**Pull request by YicunDuanUMich**
* improve the multidetection according to YOLOv10
* add some calibration plots to assess posterior distribution

### CODE REVIEW

File: bliss/catalog.py [hunk 1]
```diff
@@ -126,7 +126,7 @@ class TileCatalog(BaseTileCatalog):
     galaxy_params_index = {k: i for i, k in enumerate(galaxy_params)}
 
     def __init__(self, d: Dict[str, Tensor]):
-        self.max_sources = d["locs"].shape[3]
+        assert "locs" in d
```
Comment (jeff-regier) [inline bliss/catalog.py:129]: Could we also assert that locs is a 5d tensor here?

File: bliss/catalog.py [hunk 2]
```diff
@@ -347,12 +358,24 @@ def filter_by_flux(self, min_flux=0, max_flux=torch.inf, band=2):
...
         d = {}
+        nan_ok_variables = {
+            "shear",
+            "convergence",
+            "ellipticity",
+            "blendedness",
+            "locs_vsbc",
+            "ellipticity_vsbc",
+            "star_fluxes_vsbc",
+            "galaxy_fluxes_vsbc",
```
Comment (jeff-regier) [inline bliss/catalog.py:369]: Better not to have a list of case-study specific variable names here

Comment (jeff-regier) [reply bliss/catalog.py:369]: Is it possible to disallow nans from all variables?

File: bliss/catalog.py [hunk 3]
```diff
@@ -347,12 +358,24 @@ def filter_by_flux(self, min_flux=0, max_flux=torch.inf, band=2):
...
+        }
         for key, val in sorted_self.items():
-            if key in {"shear", "convergence"}:
```
Comment (jeff-regier) [inline bliss/catalog.py:351]: Good to remove special logic for shear and convergence here if at all possible

File: bliss/catalog.py [hunk 4]
```diff
@@ -390,6 +413,25 @@ def union(self, other, disjoint=False):
                 d[k] = torch.where(ns11 > 0, d1, d2)
         return TileCatalog(d)
 
+    def stack(self, other):
```
Comment (jeff-regier) [inline bliss/catalog.py:416]: Is this different than the merge  method?

File: bliss/encoder/variational_dist.py [hunk 5]
```diff
@@ -304,6 +305,14 @@ def mode(self):
         assert (self.base_dist.mean >= 0).all() and (self.base_dist.mean <= 1).all()
         return self.base_dist.mode
 
+    @property
+    def batch_shape(self):
+        return self.base_dist.batch_shape
+
+    @property
+    def event_shape(self):
+        return self.base_dist.event_shape
```
Comment (jeff-regier) [inline bliss/encoder/variational_dist.py:314]: OK makes sense

File: case_studies/dc2_cataloging/utils/variational_dist.py [hunk 6]
```diff
@@ -1,232 +1,117 @@
...
+        # for now, only test the vsbc for locs, ellipticity and flux
+        vsbc_variables = {
+            "locs",
+            "ellipticity",
+            "star_fluxes",
+            "galaxy_fluxes",
+        }
+        for qk, params in fp_pairs:
+            if qk.name in vsbc_variables:
+                sample_result[qk.name + "_vsbc"] = self._get_vsbc(qk, params, true_tile_cat)
```
Comment (jeff-regier) [inline case_studies/dc2_cataloging/utils/variational_dist.py:78]: Does the VSBC metric deal with checkerboard structure of the variational distribution correctly? It seems tricky to apply VSBC with the checkerboard because we can't just read off the CDF of the marginal, as we could if the variational distribution were mean field

File: case_studies/dc2_multidetection/train_config.yaml [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+    _target_: bliss.encoder.metrics.DetectionPerformance
+    bin_cutoffs: [23.9, 24.1, 24.5, 24.9, 25.6]
+    bin_type: "njymag"
+    filter_list:
+      - _target_: case_studies.dc2_cataloging.utils.metrics.SourceCountFilter
+        filter_source_count: "2plus"
+  source_type_accuracy:
+    _target_: bliss.encoder.metrics.SourceTypeAccuracy
+    bin_cutoffs: [23.9, 24.1, 24.5, 24.9, 25.6]
+    bin_type: "njymag"
```
Comment (jeff-regier) [inline case_studies/dc2_multidetection/train_config.yaml:97]: I've been reading the SDSS using AB magnitude too. (See https://classic.sdss.org/dr7/algorithms/fluxcal.php) Given that, is njymag vs mag the correct distinction? I do agree that mag has different meanings for different surveys, because the surveys bands have different filter curves, but I'm not that's captured by the njymag/mag distinction.

Comment (jeff-regier) [reply case_studies/dc2_multidetection/train_config.yaml:97]: Perhaps bliss (the main package) should be agnostic to the units of flux. Then each case study would use whatever units are appropriate for the survey.

**Comment by YicunDuanUMich** [comment]
Thanks for the review. I'll fix them tomorrow.

File: bliss/catalog.py (hunk 2 above)
Comment (YicunDuanUMich) [reply bliss/catalog.py:369]: I've removed the nan test here. nans may be necessary. They are different from 0 and play the role of mask.

I also remove the max_flux in this function. It doesn't function correctly.

File: bliss/catalog.py (hunk 4 above)
Comment (YicunDuanUMich) [reply bliss/catalog.py:416]: The key of `stack` is to add the `n_sources_mask`. This variable will be very helpful in multidetection. Since the sources in tensor are sparse after using this function, I didn't call it union.

Comment (jeff-regier) [reply bliss/catalog.py:416]: I'm reluctant to add new methods to the main bliss package that are used solely in one case study. Could you keep this function outside of the TileCatalog class, in the case study where it's used, instead?

File: bliss/catalog.py (hunk 2 above)
Comment (jeff-regier) [reply bliss/catalog.py:369]: Sounds good

File: bliss/catalog.py (hunk 4 above)
Comment (YicunDuanUMich) [reply bliss/catalog.py:416]: get it

File: bliss/catalog.py (hunk 1 above)
Comment (YicunDuanUMich) [reply bliss/catalog.py:98]: have fixed

File: bliss/catalog.py (hunk 4 above)
Comment (YicunDuanUMich) [reply bliss/catalog.py]: remove the stack

File: case_studies/dc2_cataloging/utils/variational_dist.py (hunk 6 above)
Comment (YicunDuanUMich) [reply case_studies/dc2_cataloging/utils/variational_dist.py]: I'll try to find a way to bypass it in the next PR.

File: case_studies/dc2_multidetection/train_config.yaml (hunk 7 above)
Comment (YicunDuanUMich) [reply case_studies/dc2_multidetection/train_config.yaml]: I realize that this may change the code of others' case stuides. Maybe it's better to put it in a specific PR.

**Review by jeff-regier — APPROVED** [review]
